include/libopencm3/stm32/fdcan.h

```diff
@@ -0,0 +1,793 @@
+ */
+#define FDCAN_SFEC_DISABLE				(0x0 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_FIFO0				(0x1 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_FIFO1				(0x2 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_REJECT				(0x3 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_PRIO					(0x4 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_PRIO_FIFO0			(0x5 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_PRIO_FIFO1			(0x6 << FDCAN_SFEC_SHIFT)
+/**@}*/
+
+/** Amount of standard filters allocated in Message RAM
+ * This number may vary between devices. 28 is value valid
+ * for STM32G4
+ **/
+#define FDCAN_SFT_MAX_NR				28
+
+/* SFEC = 0x7 is unused */
+
+#define FDCAN_SFID1_SHIFT				16
+#define FDCAN_SFID1_MASK				(0x7FF << FDCAN_SFID1_SHIFT)
+
+#define FDCAN_SFID2_SHIFT				0
+#define FDCAN_SFID2_MASK				(0x7FF << FDCAN_SFID2_SHIFT)
+
+/** Structure describing extended ID filters.
+ * Extended ID filter is composed of two 32bit values.
+ * This region of memory cannot be accessed in quantities less than 32bits.
+ */
+struct fdcan_extended_filter {
+	/** Aggregate of filter action and extended ID */
+	uint32_t conf_id1;
+	/** Aggregate of filter type and extended ID or mask */
+	uint32_t type_id2;
+};
+
+#define FDCAN_EFEC_SHIFT				29
+#define FDCAN_EFEC_MASK					(0x7 << FDCAN_EFEC_SHIFT)
+
+/** @defgroup fdcan_eft Extended ID filter action
+ * @{
+ */
+#define FDCAN_EFEC_DISABLE				(0x0 << FDCAN_EFEC_SHIFT)
+#define FDCAN_EFEC_FIFO0				(0x1 << FDCAN_EFEC_SHIFT)
+#define FDCAN_EFEC_FIFO1				(0x2 << FDCAN_EFEC_SHIFT)
+#define FDCAN_EFEC_REJECT				(0x3 << FDCAN_EFEC_SHIFT)
+#define FDCAN_EFEC_PRIO					(0x4 << FDCAN_EFEC_SHIFT)
+#define FDCAN_EFEC_PRIO_FIFO0			(0x5 << FDCAN_EFEC_SHIFT)
+#define FDCAN_EFEC_PRIO_FIFO1			(0x6 << FDCAN_EFEC_SHIFT)
+/**@}*/
+
+#define FDCAN_EFID1_SHIFT				0
+#define FDCAN_EFID1_MASK				(0x1FFFFFFF << FDCAN_EFID1_SHIFT)
+
+#define FDCAN_EFT_SHIFT					30
+#define FDCAN_EFT_MASK					(0x3 << FDCAN_EFT_SHIFT)
+
+/** @defgroup fdcan_eft Extended ID filter match type
+ * @{
+ */
+#define FDCAN_EFT_RANGE					(0x0 << FDCAN_EFT_SHIFT)
+#define FDCAN_EFT_DUAL					(0x1 << FDCAN_EFT_SHIFT)
+#define FDCAN_EFT_ID_MASK				(0x2 << FDCAN_EFT_SHIFT)
+#define FDCAN_EFT_RANGE_NOXIDAM			(0x3 << FDCAN_EFT_SHIFT)
+/**@}*/
+
+#define FDCAN_EFID2_SHIFT				0
+#define FDCAN_EFID2_MASK				(0x1FFFFFFF << FDCAN_EFID2_SHIFT)
+
+/** Amount of extended filters allocated in Message RAM
+ * This number may vary between devices. 8 is value valid
+ * for STM32G4
+ **/
+#define FDCAN_EFT_MAX_NR				8
+
+/** Structure describing receive FIFO element.
+ * Receive FIFO element consists of 2 32bit values for header
+ * and 16 32bit values for message payload.
+ * This area of memory can only be accessed in 32bit quantities
+ */
+struct fdcan_rx_fifo_element {
+	/** Aggregate of message identifier and flags. */
+	uint32_t identifier_flags;
+	/**< Aggregate of filter match ID, transfer format, DLC and timestamp */
+	uint32_t filt_fmt_dlc_ts;
+	/**< Message payload data */
+	uint32_t data[64 / sizeof(uint32_t)];
```

Comment:
why.... not just put 16 here?  like yu say, it has space for 16 32bit values, and it's defined as a uint32_t array.... so? just make it clear and simple?

Reply:
I chose this deliberately complicated way of defining the size, because value of 16 is actually calculated. Spec just says, that there is space for 64 data bytes allocated. It doesn't explicitly state the size of FIFO / buffer element anywhere. IMO it provides more clue to put numbers which exist in specs in, than calculate arbitrary numbers which you have to pair on your own. Especially given the fact that 64-byte long buffer is weirdly defined as an array of 32-bit ints, because byte access doesn't work with Message RAM at all.

Reply:
ok, I'm merging this now, but how abotu I add ""64bytes memory, only word accessible" or something?